pkg/inventory/inventory-client.go

```diff
@@ -6,473 +6,402 @@ package inventory
-			return nil, nil
-		}
+// String implements the fmt.Stringer interface for ID
+func (id ID) String() string {
+	return string(id)
+}
 
-		err = inv.Apply(ctx, cic.dc, cic.mapper, cic.statusPolicy)
-		return nil, err
-	}
+// Info provides the minimal information for the applier
+// to create, look up and update an inventory.
+// The inventory object can be any type, the Provider in the applier
+// needs to know how to create, look up and update it based
+// on the Info.
+type Info interface {
+	// Namespace of the inventory object.
+	// It should be the value of the field .metadata.namespace.
+	Namespace() string
 
-	// Update existing cluster inventory with merged union of objects
-	clusterObjs, err := cic.GetClusterObjs(ctx, localInv)
-	if err != nil {
-		return pruneIDs, err
-	}
-	pruneIDs = clusterObjs.Diff(objs)
-	unionObjs := clusterObjs.Union(objs)
-	var status []actuation.ObjectStatus
-	if cic.statusPolicy == StatusPolicyAll {
-		status = getObjStatus(pruneIDs, unionObjs)
-	}
-	klog.V(4).Infof("num objects to prune: %d", len(pruneIDs))
-	klog.V(4).Infof("num merged objects to store in inventory: %d", len(unionObjs))
-	wrappedInv, err := cic.InventoryFactoryFunc(clusterInv)
-	if err != nil {
-		return pruneIDs, err
-	}
-	if err = wrappedInv.Store(unionObjs, status); err != nil {
-		return pruneIDs, err
-	}
+	// ID of the inventory object.
+	// The inventory client uses this to determine how to get/update the object(s)
+	// from the cluster.
+	ID() ID
 
-	// Update not required when all objects in inventory are the same and
-	// status does not need to be updated. If status is stored, always update the
-	// inventory to store the latest status.
-	if objs.Equal(clusterObjs) && cic.statusPolicy == StatusPolicyNone {
-		return pruneIDs, nil
-	}
+	GetLabels() map[string]string
 
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infof("dry-run create inventory object: not created")
-		return pruneIDs, nil
-	}
-	err = wrappedInv.Apply(ctx, cic.dc, cic.mapper, cic.statusPolicy)
-	return pruneIDs, err
+	GetAnnotations() map[string]string
 }
 
-// Replace stores the passed objects in the cluster inventory object, or
-// an error if one occurred.
-func (cic *ClusterClient) Replace(ctx context.Context, localInv Info, objs object.ObjMetadataSet, status []actuation.ObjectStatus,
-	dryRun common.DryRunStrategy) error {
-	// Skip entire function for dry-run.
-	if dryRun.ClientOrServerDryRun() {
-		klog.V(4).Infoln("dry-run replace inventory object: not applied")
-		return nil
-	}
-	clusterInv, err := cic.getClusterInventoryInfo(ctx, localInv)
-	if err != nil {
-		return fmt.Errorf("failed to read inventory from cluster: %w", err)
-	}
+type SimpleInfo struct {
+	id          ID
+	namespace   string
+	labels      map[string]string
+	annotations map[string]string
+}
 
-	clusterObjs, err := cic.GetClusterObjs(ctx, localInv)
-	if err != nil {
-		return fmt.Errorf("failed to read inventory objects from cluster: %w", err)
-	}
+func (i SimpleInfo) ID() ID {
```

Comment:
Lets change these to GetID and GetNamespace, just to avoid needing to prefix the field names in the fakes. If we expose the fields on SimpleInfo & SingleObjectInfo we might not need the fakes at all.